001-finding_resources.md

```diff
@@ -0,0 +1,120 @@
+finding other resources, such as plugins.
+
+### Current Process
+
+This section describes how Ignition Gazebo currently finds Mesh URI
+resources. SDF contains multiple instances of URI elements, but the current
+implementation of Gazebo handles only mesh URIs. The information contained
+here can be extrapolated to handle other types of resources, such as
+textures and audio files.
+
+A user of Ignition may not need or use Gazebo. If this describes you, then
+skip directly to the last step which describes how
+`ignition::common::findFile` can be used to resolve a resource.
+
+1. Ignition Gazebo registers an SDF find callback in the Server constructor.
+   This callback is triggered by SDF whenever a URI resource from an
+   `<include>` element cannot be found. The Ignition Gazebo find callback in
+   turn calls Fuel Tools to download the resource.
+
+2. After registering the callback, Ignition Gazebo then finds the world file
+   specified on the command line in a path contained in the
+   `IGN_GAZEBO_RESOURCE_PATH` environment variable using
+   `common::SystemPaths::FindFile`.
+
+3. If the world file is found, then SDF is used to load the world file. SDF
+   will only attempt to resolve URI resources contained withing the world
+   file that are in `<include>` elements.
+
+4. A SimulationRunner is then instantiated for each `sdf::World`. The
+   LevelManager, contained in each SimulationRunner, loads each `sdf::Model`
+   and creates the appropriate set of components. Plugin loading is conducted
+   during this step, but no mesh URI resource loading is conducted.
+
+5. Systems and other aspects of simulation, such as rendering, must then each
+   resolve URI resources contained within the loaded world and models.
+   The `Util::asFullPath` is a helper function that is used to convert a URI
+   into a system path. This helper function only alters URIs that have
+   a relative path. URIs that have a scheme or are a full path are
+   untouched.
+
+6. The result of `Util::asFullPath` is then passed to
+   `ignition::common::MeshManager::Load`, which then calls
+   `ignition::common::findFile` to finally resolve the location of the URI
+   resource.
+
+7. `ignition::common::findFile` executes the following procedure.
+
+    1. Calls `SystemPaths::FindFile(string file, bool searchLocalPath)`
+    2. `SystemPaths::FindFile(string file, bool searchLocalPath)` executes:
+        1. if `file` is empty, then return an empty string
+        2. if `file` is a valid URI, then `path`
+           = `SystemsPath::FindFileURI(file)`. The `FindFileURI(file)`
+           executes:
+            1. if the URI scheme is "file", then `path` = `SystemsPath::FindFile(_file)`. This will only be executed if `SystemsPath::FindFileURI(file)` is called indepently of `SystemPaths::FindFile`.
+            2. else if there is a user define CB, the `path` = `FileURICB(_file)`. This `FileURICB` is deprecated in ign-common3.
```

Comment:
I believe it's deprecated not because callbacks are no longer supported, but because users should use `AddFindFile*Callback` instead of `Set...`. So maybe this document could just mention the `Add` API and leave out the comment about deprecation?

Reply:
Clarified this in a3696f8e717265fb69c906bc630f83723b50d8dd